File: backend/app/schemas/attendance.py
"""Pydantic schemas for attendance endpoints."""
from typing import List, Optional
from pydantic import BaseModel


class AttendanceRecordOut(BaseModel):
    id: str
    student_id: str
    class_id: str
    date: str          # YYYY-MM-DD
    status: str        # present | absent | late | half_day
    remarks: Optional[str] = None

    model_config = {"from_attributes": True}


class AttendanceSummaryOut(BaseModel):
    student_id: str
    total_days: int
    present_days: int
    absent_days: int
    late_days: int
    percentage: float


class MarkAttendanceItem(BaseModel):
    student_id: str
    status: str        # present | absent | late | half_day
    remarks: Optional[str] = None


class MarkAttendanceRequest(BaseModel):
    class_id: str
    date: str          # YYYY-MM-DD
    records: List[MarkAttendanceItem]

File: src/services/attendance.ts
import { api } from './api';

export interface AttendanceSummary {
  student_id: string;
  total_days: number;
  present_days: number;
  absent_days: number;
  late_days: number;
  percentage: number;
}

export interface AttendanceRecord {
  id: string;
  student_id: string;
  class_id: string;
  date: string;
  status: 'present' | 'absent' | 'late' | 'half_day';
  remarks?: string;
}

export interface MarkAttendanceItem {
  student_id: string;
  status: string;
  remarks?: string;
}

export async function getAttendanceSummary(studentId: string): Promise<AttendanceSummary> {
  return api.get<AttendanceSummary>(`/attendance/student/${studentId}/summary`);
}

export async function getAttendanceRecords(studentId: string, limit = 30): Promise<AttendanceRecord[]> {
  return api.get<AttendanceRecord[]>(`/attendance/student/${studentId}/records?limit=${limit}`);
}

export async function markAttendance(classId: string, date: string, records: MarkAttendanceItem[]): Promise<{ marked: number; date: string }> {
  return api.post(`/attendance/class/${classId}/mark`, { class_id: classId, date, records });
}
